internal/provider/resource_controller.go

```diff
@@ -0,0 +1,770 @@
+	Name           types.String `tfsdk:"name"`
+	AuthType       types.Set    `tfsdk:"auth_type"`
+	CACertificates types.Set    `tfsdk:"ca_certificates"`
+	Config         types.Map    `tfsdk:"config"`
+	Endpoint       types.String `tfsdk:"endpoint"`
+	Region         types.Object `tfsdk:"region"`
+	Type           types.String `tfsdk:"type"`
+}
+
+// nestedCloudRegionModel represents the region nested object in the cloud
+type nestedCloudRegionModel struct {
+	Name             types.String `tfsdk:"name"`
+	Endpoint         types.String `tfsdk:"endpoint"`
+	IdentityEndpoint types.String `tfsdk:"identity_endpoint"`
+	StorageEndpoint  types.String `tfsdk:"storage_endpoint"`
+}
+
+// nestedCloudCredentialModel represents cloud credentials configuration
+type nestedCloudCredentialModel struct {
+	Name       types.String `tfsdk:"name"`
+	AuthType   types.String `tfsdk:"auth_type"`
+	Attributes types.Map    `tfsdk:"attributes"`
+}
+
+// controllerConnectionInformation contains the connection details for a controller.
+type controllerConnectionInformation struct {
+	Addresses []string
+	CACert    string
+	Username  string
+	Password  string
+}
+
+type boostrapArguments struct {
+	AgentVersion         string
+	BootstrapBase        string
+	BootstrapConstraints map[string]string
+	Cloud                bootstrapCloudArgument
+	CloudCredential      bootstrapCredentialArgument
+	Config               map[string]string
+	JujuBinary           string
+	ModelConstraints     map[string]string
+	ModelDefault         map[string]string
+	Name                 string
+	NumUnits             int64
+	OutputFile           string
+	StoragePool          map[string]string
+}
+
+type bootstrapCloudArgument struct {
+	Name           string
+	AuthTypes      []string
+	CACertificates []string
+	Config         map[string]string
+	Endpoint       string
+	Region         *boostrapCloudRegionArgument
+	Type           string
+}
+
+type boostrapCloudRegionArgument struct {
+	Name             string
+	Endpoint         string
+	IdentityEndpoint string
+	StorageEndpoint  string
+}
+
+type bootstrapCredentialArgument struct {
+	Name       string
+	AuthType   string
+	Attributes map[string]string
+}
+
+type jujuCommand interface {
+	// Bootstrap creates a new controller and returns connection information.
+	Bootstrap(ctx context.Context, model boostrapArguments) (*controllerConnectionInformation, error)
+	// UpdateConfig updates controller configuration.
+	UpdateConfig(ctx context.Context, connInfo *controllerConnectionInformation, config map[string]string) error
+	// Config retrieves controller configuration settings.
+	Config(ctx context.Context, connInfo *controllerConnectionInformation) (map[string]string, error)
+	// Destroy removes the controller.
+	Destroy(ctx context.Context, connInfo *controllerConnectionInformation) error
+}
+
+var defaultNewJujuCommandFunction func(string) jujuCommand = func(jujuBinary string) jujuCommand {
+	return &defaultJujuCommand{jujuBinary: jujuBinary}
+}
```

Comment:
I think all of this needs to go inside `internal/juju`, to provide a separation between provider logic and our domain logic.

Reply:
moved